### PR TITLE
Fix service session affinity e2e failure cases

### PR DESCRIPTION
**Pull request by m1093782566**
**What this PR does / why we need it**:

Fix service session affinity e2e failure cases - debuging...

**Which issue this PR fixes**:

xref #54571 #54524

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

/sig network

### CODE REVIEW

File: test/e2e/framework/networking_utils.go [hunk 1]
```diff
@@ -220,16 +220,16 @@ func (config *NetworkingTestConfig) DialFromContainer(protocol, containerIP, tar
...
-// GetEndpointsFromContainer executes a curl via kubectl exec in a test container,
+// GetDiffEndpointsFromContainer executes a curl via kubectl exec in a test container,
 // which might then translate to a tcp or udp request based on the protocol argument
-// in the url.
-// - tries is the maximum number of curl attempts. If this many attempts pass and
+// in the url. It returns all different endpoints from multiple retries.
+// - tries is the number of curl attempts. If this many attempts pass and
 //   we don't see any endpoints, the test fails.
-func (config *NetworkingTestConfig) GetEndpointsFromContainer(protocol, containerIP, targetIP string, containerHttpPort, targetPort, tries int) (sets.String, error) {
+func (config *NetworkingTestConfig) GetDiffEndpointsFromContiner(protocol, containerIP, targetIP string, containerHttpPort, targetPort, tries int) (sets.String, error) {
```
Comment (cmluciano) [inline test/e2e/framework/networking_utils.go]: typo 

Continer

File: test/e2e/framework/networking_utils.go [hunk 2]
```diff
@@ -261,15 +261,14 @@ func (config *NetworkingTestConfig) GetEndpointsFromContainer(protocol, containe
...
-			if len(eps) > 0 {
-				return eps, nil
-			}
 			// TODO: get rid of this delay #36281
 			time.Sleep(hitEndpointRetryDelay)
 		}
 	}
-	return nil, fmt.Errorf("error getting endpoints:\nTries %d\nCommand %v\n", tries, cmd)
+	if len(eps) == 0 {
+		return nil, fmt.Errorf("Unexpetced no endpoints return")
```
Comment (cmluciano) [inline test/e2e/framework/networking_utils.go]: typo 

Unexpected

**Review by cmluciano — CHANGES_REQUESTED** [review]
some typos

File: test/e2e/network/networking.go [hunk 3]
```diff
@@ -98,6 +98,54 @@ var _ = SIGDescribe("Networking", func() {
...
+			framework.Failf("Failed to update service session affinity, error: %v", err)
+		}
+
+		// debug
+		framework.Logf("Updated service: %v", config.NodePortService)
+
+		// Check if number of endpoints returned are exactly one.
+		eps, err := config.GetDiffEndpointsFromTestContainer("udp", config.ClusterIP, framework.ClusterUdpPort, framework.SessionAffinityChecks)
+		if err != nil {
+			framework.Failf("Unexpetced error: %v", err)
```
Comment (spiffxp) [inline test/e2e/network/networking.go]: Would prefer to explain when/why, eg:

"Failed to get endpoints from test container, error: "

File: test/e2e/network/networking.go [hunk 4]
```diff
@@ -98,6 +98,54 @@ var _ = SIGDescribe("Networking", func() {
...
+		}
+		if len(eps) > 1 {
+			framework.Failf("Unexpetced endpoints return: %v, expect 1 endpoints", eps)
```
Comment (spiffxp) [inline test/e2e/network/networking.go]: typo, Unexpected

File: test/e2e/framework/networking_utils.go (hunk 1 above)
Comment (spiffxp) [inline test/e2e/framework/networking_utils.go]: Why the rename? Diff to me implies comparing or a delta between an old and new.  I don't think it's doing that here?  As a name, `GetEndpointsFromContainer` already tells me it's getting all of the container's endpoints?

File: test/e2e/network/networking.go [hunk 5]
```diff
@@ -98,6 +98,54 @@ var _ = SIGDescribe("Networking", func() {
...
+			framework.Failf("Failed to update service session affinity, error: %v", err)
+		}
+
+		// debug
+		framework.Logf("Updated service: %v", config.NodePortService)
+
+		// Check if number of endpoints returned are exactly one.
+		eps, err := config.GetDiffEndpointsFromTestContainer("http", config.ClusterIP, framework.ClusterHttpPort, framework.SessionAffinityChecks)
+		if err != nil {
+			framework.Failf("Unexpetced error: %v", err)
```
Comment (spiffxp) [inline test/e2e/network/networking.go]: Would prefer to explain when/why, eg:

"Failed to get endpoints from test container, error: "

File: test/e2e/network/networking.go [hunk 6]
```diff
@@ -98,6 +98,54 @@ var _ = SIGDescribe("Networking", func() {
...
+		}
+		if len(eps) > 1 {
+			framework.Failf("Unexpetced endpoints return: %v, expect 1 endpoints", eps)
```
Comment (spiffxp) [inline test/e2e/network/networking.go]: typo, Unexpected

File: test/e2e/framework/networking_utils.go (hunk 1 above)
Comment (spiffxp) [reply test/e2e/framework/networking_utils.go]: I guess it looks like the original impl of this was buggy and returned too soon? Still think the name fits the intent here

**Review by spiffxp — COMMENTED** [review]
a few nits

File: test/e2e/framework/networking_utils.go (hunk 2 above)
Comment (MrHohn) [inline test/e2e/framework/networking_utils.go]: I think this should be handled outside of `GetEndpointsFromContainer()`.

File: test/e2e/framework/networking_utils.go (hunk 1 above)
Comment (MrHohn) [reply test/e2e/framework/networking_utils.go]: +1, `GetEndpointsFromContainer` seems to be a proper name.

File: test/e2e/framework/networking_utils.go [hunk 7]
```diff
@@ -247,7 +247,7 @@ func (config *NetworkingTestConfig) GetEndpointsFromContainer(protocol, containe
 			// we confirm unreachability.
 			Logf("Failed to execute %q: %v, stdout: %q, stderr: %q", cmd, err, stdout, stderr)
 		} else {
-			Logf("maxTries: %d, in try: %d, stdout: %v, stderr: %v", tries, i, stdout, stderr)
+			Logf("maxTries: %d, in try: %d, stdout: %v, stderr: %v, command run in: %#v", tries, i, stdout, stderr, config.HostTestContainerPod)
```
Comment (MrHohn) [inline test/e2e/framework/networking_utils.go]: s/maxTries/tries

**Comment by m1093782566** [comment]
We can see 

[sig-network] Networking should function for client IP based session affinity: http
[sig-network] Networking should function for client IP based session affinity: udp

is passing in `pull-kubernetes-e2e-gce`

Should we trigger it more to see if it's still flaky?

**Comment by cmluciano** [comment]
/test pull-kubernetes-e2e-gce

triggering one more time

**Review by cmluciano — APPROVED** [review]

File: test/e2e/network/networking.go [hunk 8]
```diff
@@ -98,6 +98,60 @@ var _ = SIGDescribe("Networking", func() {
...
+		updateSessionAffinity := func(svc *v1.Service) {
+			svc.Spec.SessionAffinity = v1.ServiceAffinityClientIP
+		}
+		_, err := framework.UpdateService(f.ClientSet, config.NodePortService.Namespace, config.NodePortService.Name, updateSessionAffinity)
+		if err != nil {
+			framework.Failf("Failed to update service session affinity, error: %v", err)
+		}
+
+		// debug
+		framework.Logf("Updated service: %v", config.NodePortService)
```
Comment (MrHohn) [inline test/e2e/network/networking.go]: Remove debug logging.

File: test/e2e/network/networking.go [hunk 9]
```diff
@@ -98,6 +98,60 @@ var _ = SIGDescribe("Networking", func() {
...
```
Comment (MrHohn) [inline test/e2e/network/networking.go]: Remove debug logging.

**Review by MrHohn — COMMENTED** [review]
/hold

We need to put the tests back into slow suite before merging.

File: test/e2e/framework/networking_utils.go [hunk 10]
```diff
@@ -261,15 +261,11 @@ func (config *NetworkingTestConfig) GetEndpointsFromContainer(protocol, containe
...
-			// Return immediately when we successfully fetch endpoints
-			if len(eps) > 0 {
-				return eps, nil
-			}
 			// TODO: get rid of this delay #36281
 			time.Sleep(hitEndpointRetryDelay)
 		}
 	}
-	return nil, fmt.Errorf("error getting endpoints:\nTries %d\nCommand %v\n", tries, cmd)
+	return eps, nil
```
Comment (MrHohn) [inline test/e2e/framework/networking_utils.go]: Humm, why shouldn't we return error here?

**Review by MrHohn — COMMENTED** [review]
/lgtm
/hold cancel

File: test/e2e/framework/networking_utils.go (hunk 10 above)
Comment (MrHohn) [reply test/e2e/framework/networking_utils.go]: Never mind, this seems correct :)

**Pull request merged**